x/stake/keeper/slash.go

```diff
@@ -145,34 +145,42 @@ func (k Keeper) Unjail(ctx sdk.Context, consAddr sdk.ConsAddress) {
 // (the amount actually slashed may be less if there's
 // insufficient stake remaining)
 func (k Keeper) slashUnbondingDelegation(ctx sdk.Context, unbondingDelegation types.UnbondingDelegation,
-	infractionHeight int64, slashFactor sdk.Dec) (slashAmount sdk.Int) {
+	infractionHeight int64, slashFactor sdk.Dec) (totalSlashAmount sdk.Int) {
 
 	now := ctx.BlockHeader().Time
+	totalSlashAmount = sdk.ZeroInt()
 
-	// If unbonding started before this height, stake didn't contribute to infraction
-	if unbondingDelegation.CreationHeight < infractionHeight {
-		return sdk.ZeroInt()
-	}
+	// perform slashing on all entries within the unbonding delegation
+	for i, entry := range unbondingDelegation.Entries {
 
-	if unbondingDelegation.MinTime.Before(now) {
-		// Unbonding delegation no longer eligible for slashing, skip it
-		// TODO Settle and delete it automatically?
-		return sdk.ZeroInt()
-	}
+		// If unbonding started before this height, stake didn't contribute to infraction
+		if entry.CreationHeight < infractionHeight {
+			continue
+		}
 
-	// Calculate slash amount proportional to stake contributing to infraction
-	slashAmountDec := slashFactor.MulInt(unbondingDelegation.InitialBalance.Amount)
-	slashAmount = slashAmountDec.TruncateInt()
+		if entry.IsMature(now) {
+			// Unbonding delegation no longer eligible for slashing, skip it
+			// TODO Settle and delete it automatically?
+			continue
+		}
 
-	// Don't slash more tokens than held
-	// Possible since the unbonding delegation may already
-	// have been slashed, and slash amounts are calculated
-	// according to stake held at time of infraction
-	unbondingSlashAmount := sdk.MinInt(slashAmount, unbondingDelegation.Balance.Amount)
+		// Calculate slash amount proportional to stake contributing to infraction
+		slashAmountDec := slashFactor.MulInt(entry.InitialBalance.Amount)
+		slashAmount := slashAmountDec.TruncateInt()
+		totalSlashAmount = totalSlashAmount.Add(slashAmount)
 
-	// Update unbonding delegation if necessary
-	if !unbondingSlashAmount.IsZero() {
-		unbondingDelegation.Balance.Amount = unbondingDelegation.Balance.Amount.Sub(unbondingSlashAmount)
+		// Don't slash more tokens than held
+		// Possible since the unbonding delegation may already
+		// have been slashed, and slash amounts are calculated
+		// according to stake held at time of infraction
+		unbondingSlashAmount := sdk.MinInt(slashAmount, entry.Balance.Amount)
+
+		// Update unbonding delegation if necessary
+		if unbondingSlashAmount.IsZero() {
+			continue
+		}
+		entry.Balance.Amount = entry.Balance.Amount.Sub(unbondingSlashAmount)
+		unbondingDelegation.Entries[i] = entry
 		k.SetUnbondingDelegation(ctx, unbondingDelegation)
```

Comment:
This is called even if no unbonding delegations were slashed, which isn't necessary

Reply:
that's incorrect, functionally the code has stayed the same as the if statement which this line used to be enclosed is simply a continue statement above (with reverse condition)

Reply:
Ah nvm misread the diff, just checked the file and all looks :100: